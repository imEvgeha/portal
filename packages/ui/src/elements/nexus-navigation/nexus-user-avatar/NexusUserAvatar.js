import React, {useState} from 'react';
import PropTypes from 'prop-types';
import LogoutIcon from '@vubiquity-nexus/portal-assets/logout.svg';
import TenantIcon from '@vubiquity-nexus/portal-assets/tenant.svg';
import {logout} from '@vubiquity-nexus/portal-auth/authActions';
import {Button} from 'primereact/button';
import {Divider} from 'primereact/divider';
import {Dropdown} from 'primereact/dropdown';
import {connect} from 'react-redux';
import './NexusUserAvatar.scss';

/**
 *
 * @param {profileInfo} Retrieves User profile from Redux
 * @param {logout} Dispatch function to logout the user from session
 * @returns NexusUserAvatar
 */
const NexusUserAvatar = ({profileInfo, logout}) => {
    const [selectedTenant, setSelectedTenant] = useState(null);

    const tenants = [
        {name: 'WB', code: 'WB'},
        {name: 'MGM', code: 'MGM'},
        {name: 'Vubiquity', code: 'VU'},
    ];

    /**
     *
     * @param {option} option Selected option from the dropdown
     * @param {props} props props of the dropdown
     * @returns
     */
    const selectedTenantTemplate = (option, props) => {
        if (option) {
            return (
                <div className="country-item country-item-value">
                    <img
                        alt={option.name}
                        src={TenantIcon}
                        onError={e =>
                            (e.target.src = 'https://www.primefaces.org/wp-content/uploads/2020/05/placeholder.png')
                        }
                        className={`flag flag-${option.code.toLowerCase()}`}
                    />
                    <div>{option.name}</div>
                </div>
            );
        }

        // return <span>{props.placeholder}</span>;
    };

    const tenantOptionTemplate = option => {
        return (
            <div className="country-item">
                <img
                    alt={option.name}
                    src={TenantIcon}
                    onError={e =>
                        (e.target.src = 'https://www.primefaces.org/wp-content/uploads/2020/05/placeholder.png')
                    }
                    className={`flag flag-${option.code.toLowerCase()}`}
                />
                <div>{option.name}</div>
            </div>
        );
    };

    const onTenantChange = e => {
        setSelectedTenant(e.value);
    };

    return (
        <div className="NexusUserAvatar">
            <div className="UserAvatarHeading">
                <div className="UserAvatarFullName">
                    {`${profileInfo.firstName} ${profileInfo.lastName}` || 'Profile'}
                </div>
                <div className="UserAvatarUsername">{`${profileInfo.username}`}</div>
            </div>
            <Divider />
            <div className="UserAvatarTenantsSelection">
                <Dropdown
                    value={selectedTenant}
                    options={tenants}
                    onChange={e => onTenantChange(e.value)}
                    optionLabel="name"
                    placeholder="Select a Tenant"
                    valueTemplate={selectedTenantTemplate}
                    itemTemplate={tenantOptionTemplate}
                    dropdownIcon="pi pi-chevron-right"
                    showOnFocus
                />
            </div>
            <Divider />
            <div className="UserAvatarLogout">
                <Button className="logout p-0" aria-label="LogOut" onClick={logout}>
                    <i className={`${LogoutIcon} px-2`} />
                    <span className="px-3">Log Out</span>
                </Button>
            </div>
        </div>
    );
};

NexusUserAvatar.propTypes = {
    profileInfo: PropTypes.object,
    logout: PropTypes.func,
};

NexusUserAvatar.defaultProps = {
    profileInfo: {},
    logout: () => null,
};

const mapStateToProps = ({auth}) => {
    const {userAccount} = auth || {};
    return {
        profileInfo: userAccount,
    };
};

const mapDispatchToProps = dispatch => ({
    logout: payload => dispatch(logout(payload)),
});

export default connect(mapStateToProps, mapDispatchToProps)(NexusUserAvatar);
