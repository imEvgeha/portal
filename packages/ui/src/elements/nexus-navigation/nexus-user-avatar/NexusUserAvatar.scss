.NexusUserAvatar {
    padding: 20px;
    color: black;
    .UserAvatarHeading {
        .UserAvatarFullName {
            font: normal normal bold 20px/23px Arial;
        }
        .UserAvatarUsername {
            margin-top: 2px;
            font: normal normal normal 12px/14px Arial;
        }
    }
    .UserAvatarTenantsSelection {
    }
    .NexusUserAvatar .UserAvatarLogout .p-button.logout {
        background: linear-gradient(to left, var(--bluegray-700) 50%, var(--bluegray-800) 50%);
        background-size: 200% 100%;
        background-position: right bottom;
        transition: background-position 0.5s ease-out;
        color: rgb(0, 0, 0);
        border-color: var(--bluegray-800);
    }
    .NexusUserAvatar .UserAvatarLogout .p-button.logout:hover {
        background-position: left bottom;
    }
    .NexusUserAvatar .UserAvatarLogout .p-button.logout i {
        background-color: var(--bluegray-800);
    }
    .NexusUserAvatar .UserAvatarLogout .p-button.logout:focus {
        box-shadow: 0 0 0 1px var(--bluegray-500);
    }
}
