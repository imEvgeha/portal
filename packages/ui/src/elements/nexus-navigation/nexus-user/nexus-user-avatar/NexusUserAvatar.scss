.NexusUserAvatar {
    padding: 8px;
    color: black;
    .UserAvatarHeading {
        .UserAvatarFullName {
            font: normal normal bold 20px/15px Arial;
        }
        .UserAvatarUsername {
            margin-top: 4px;
            font: normal normal normal 12px/18px Arial;
        }
        .ApplicationVersion {
            font: normal normal normal 12px/20px Arial;
        }
    }

    .p-divider-horizontal.NexusUserDivider {
        margin: 0.8rem 0;
    }

    .UserAvatarLogout {
        padding: 8px;
        .logoutIcon {
            margin-right: 5px;
        }
        .logoutButtonText {
            color: black;
        }
        .p-button.logout {
            background: rgb(0 0 0 / 0%);
            width: 100%;
            border: 0;
            padding: 8px;
        }
        .p-button.logout:hover {
            background-color: rgba(128, 125, 125, 0.103);
        }
    }
}
