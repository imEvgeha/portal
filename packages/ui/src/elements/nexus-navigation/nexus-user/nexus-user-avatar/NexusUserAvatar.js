import React from 'react';
import PropTypes from 'prop-types';
import {keycloak} from '@portal/portal-auth';
import {logout, setSelectedTenantInfo} from '@portal/portal-auth/authActions';
import {transformSelectTenant, updateLocalStorageWithSelectedTenant} from '@portal/portal-auth/utils';
import ExpandRightIcon from '@vubiquity-nexus/portal-assets/expand_right.svg';
import LogoutIcon from '@vubiquity-nexus/portal-assets/logout.svg';
import TenantIcon from '@vubiquity-nexus/portal-assets/tenant.svg';
import {getConfig} from '@vubiquity-nexus/portal-utils/lib/config';
import {Divider} from 'primereact/divider';
import {TieredMenu} from 'primereact/tieredmenu';
import {connect, useDispatch, useSelector} from 'react-redux';
import './NexusUserAvatar.scss';

/**
 *
 * @param {Object} profileInfo Redux - auth.profileInfo
 * @param {logout} logout Dispatch function to logout the user from session
 * @returns NexusUserAvatar
 */
const NexusUserAvatar = ({selectedTenant, profileInfo, logout, menu}) => {
    const dispatch = useDispatch();
    const currentLoggedInUsername = useSelector(state => state?.auth?.userAccount?.username);
    // get client roles from keycloak
    const {resourceAccess, realmAccess} = keycloak;
    // filter out clients that are not tenants
    const filteredResourceAccess = {...resourceAccess};
    delete filteredResourceAccess['account'];
    delete filteredResourceAccess['realm-management'];
    /**
     * Template used for the selected Tenant in TieredMenu
     * @param {*} item Client item
     * @param {*} options All options
     * @returns
     */
    const tenantSelectedTemplated = (item, options) => {
        return (
            <div className="tenant-selected-item" onClick={options.onClick}>
                <div className="tenantDescription">
                    <TenantIcon className="tenantIcon" />
                    <span className="tenantName">{item.label}</span>
                </div>
                <div className="actionIcon">
                    {/* Do not render an icon if there are not more than one tenant */}
                    {Object.entries(filteredResourceAccess).length > 1 && <ExpandRightIcon className="expandIcon" />}
                </div>
            </div>
        );
    };

    /**
     * Template used for all the Tenant Options in TieredMenu
     * @param {*} option Client option
     * @param {*} index Index of the client
     * @returns
     */
    const tenantOptionTemplate = (option, index) => {
        return (
            <div className="tenant-option-item" onClick={e => onTenantChange(option)}>
                <div className="tenantName">{option[0]}</div>
                {renderDivider(index) && <Divider className="TenantSelectionDivider" />}
            </div>
        );
    };

    /**
     * This is used to calculate when to render a Divider after a tenant
     * If it is the last tenant, do not render a <Divider />
     * @param {*} index Index of the tenant
     * @returns
     */
    const renderDivider = index => {
        return index + 1 < Object.keys(filteredResourceAccess).length;
    };

    /**
     * onChange function when selecting a tenant from the Dropdown list
     * transforms key:value array to object
     * updated Redux with the selected tenant for this session
     * update LocalStorage with this user's selected tenant
     * @param {array} selectedTenant - [0] - id, [1] - roles[]
     */
    const onTenantChange = selectedTenant => {
        const tempSelectedTenant = transformSelectTenant(selectedTenant);
        tempSelectedTenant.roles = [...tempSelectedTenant.roles, ...realmAccess.roles];
        dispatch(setSelectedTenantInfo(tempSelectedTenant));
        updateLocalStorageWithSelectedTenant(currentLoggedInUsername, tempSelectedTenant);
    };

    /**
     * Contruct the dynamic tenants, based on the resourceAccess from keycloak
     * Reference: https://www.primefaces.org/primereact/tieredmenu/
     */
    const tenants = () => {
        // selected tenant, parent item
        const tmpTenants = {
            label: selectedTenant.id,
            template: tenantSelectedTemplated,
            roles: selectedTenant.roles,
        };

        // sub items, tenants list
        // check if there are more tenants to be add in a list, or just one tenant
        if (Object.entries(filteredResourceAccess).length > 1) {
            tmpTenants.items = [
                ...Object.entries(filteredResourceAccess).map((client, index) => {
                    return {
                        id: client[0],
                        roles: [...client[1].roles],
                        template: tenantOptionTemplate(client, index),
                        label: client[0],
                    };
                }),
            ];
        }

        return tmpTenants;
    };

    /**
     *
     * @returns Render Header of Avatar
     */
    const RenderAvatarHeading = () => {
        return (
            <div className="UserAvatarHeading">
                <div className="UserAvatarFullName">
                    {`${profileInfo.firstName} ${profileInfo.lastName}` || 'Profile'}
                </div>
                <div className="UserAvatarUsername">{`${profileInfo.username}`}</div>
                <div className="ApplicationVersion">{`v${getConfig('portalVersion')}`}</div>
            </div>
        );
    };

    /**
     *
     * @returns Avatar Actions - Bottom section for actions
     */
    const RenderAvatarActions = () => {
        return (
            <div className="UserAvatarActions">
                <div className="tenant-logout-action" onClick={logout}>
                    <LogoutIcon className="logoutIcon" />
                    <span className="logoutButtonText">Log Out</span>
                </div>
            </div>
        );
    };

    /**
     * Contruct the items model as expected by TieredMenu
     * Reference: https://www.primefaces.org/primereact/tieredmenu/
     */
    const items = [
        {
            template: RenderAvatarHeading,
        },
        {
            separator: true,
        },
        tenants(),
        {
            separator: true,
        },
        {
            template: RenderAvatarActions,
        },
    ];

    return (
        <div className="NexusUserAvatar">
            <TieredMenu model={items} className="TenantSelection" popup ref={menu} />
        </div>
    );
};

NexusUserAvatar.propTypes = {
    selectedTenant: PropTypes.object,
    profileInfo: PropTypes.object,
    logout: PropTypes.func,
    menu: PropTypes.any,
};

NexusUserAvatar.defaultProps = {
    profileInfo: {},
    logout: () => null,
    menu: null,
    selectedTenant: {},
};

const mapStateToProps = ({auth}) => {
    const {userAccount, selectedTenant} = auth || {};
    return {
        profileInfo: userAccount,
        selectedTenant,
    };
};

const mapDispatchToProps = dispatch => ({
    logout: payload => dispatch(logout(payload)),
});

export default connect(mapStateToProps, mapDispatchToProps)(NexusUserAvatar);
