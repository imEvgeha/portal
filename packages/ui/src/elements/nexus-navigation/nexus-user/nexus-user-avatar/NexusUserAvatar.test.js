import React from 'react';
import {shallow} from 'enzyme';
import configureStore from 'redux-mock-store';
import NexusUserAvatar from './NexusUserAvatar';

describe('NexusUserAvatar', () => {
    let wrapper = null;
    let mockStore = null;
    let store = null;

    beforeEach(() => {
        mockStore = configureStore();
        // add to the mockStore the attributes necessary by this component
        store = mockStore({
            auth: {
                userAccount: {
                    username: 'mcharalambous',
                    firstName: 'Marios',
                    lastName: 'Charalambous',
                    email: 'marios.charalambous@vubiquity.net',
                    emailVerified: false,
                },
            },
        });
        wrapper = shallow(<NexusUserAvatar store={store} />);
    });

    it('should match snapshot', () => {
        expect(wrapper).toMatchSnapshot();
    });
});
