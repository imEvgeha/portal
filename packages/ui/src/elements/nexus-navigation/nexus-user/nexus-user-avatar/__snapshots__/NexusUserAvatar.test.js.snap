// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`NexusUserAvatar should match snapshot 1`] = `
<ContextProvider
  value={null}
>
  <NexusUserAvatar
    logout={[Function]}
    menu={null}
    profileInfo={
      Object {
        "email": "marios.charalambous@vubiquity.net",
        "emailVerified": false,
        "firstName": "Marios",
        "lastName": "Charalambous",
        "username": "mcharalambous",
      }
    }
    selectedTenant={Object {}}
    showTenantSelectionDropdown={false}
    store={
      Object {
        "clearActions": [Function],
        "dispatch": [Function],
        "getActions": [Function],
        "getState": [Function],
        "replaceReducer": [Function],
        "subscribe": [Function],
      }
    }
  />
</ContextProvider>
`;
